src/HelixServer.js

```diff
@@ -102,20 +111,28 @@ class HelixServer {
           .then(boundResolver)
           .then(executeTemplate)
           .then((result) => {
+            if (!result) {
+              throw new Error('Response is empty, don\'t know what to do');
+            }
             if (result instanceof Error) {
               // full response is an error: engine error
               throw result;
             }
             if (result && result.error && result.error instanceof Error) {
               throw result.error;
             }
-            if (!result || !result.body) {
-              // empty body: nothing to render
-              throw new Error('Response has no body, don\'t know what to do');
-            }
+            let body = result.body || '';
+            const headers = result.headers || {};
             const status = result.statusCode || 200;
-            esi.process(result.body).then((body) => {
-              res.status(status).send(body);
+            const contentType = headers['Content-Type'] || 'text/html';
+            if (/.*\/json/.test(contentType)) {
+              body = JSON.stringify(body);
+            } else if (/.*\/octet-stream/.test(contentType) || /image\/.*/.test(contentType)) {
```

Comment:
We should at some point probably build a centralized type-check library that detects text/json/binary. We have this check now scattered across a number of places.